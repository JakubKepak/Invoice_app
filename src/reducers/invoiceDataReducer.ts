/* eslint-disable no-case-declarations */
import { FilterType, InvoiceType } from 'types/types';

type Action =
  | { type: 'addFilter'; payload: FilterType }
  | { type: 'removeFilter'; payload: FilterType }
  | { type: 'error' }
  | { type: 'refresh'; payload: any };

export const invoiceDataReducer = (state: any, action: Action): any => {
  switch (action.type) {
    case 'refresh':
      if (action.payload) {
        const filteredData = action.payload.data.filter(
          (invoice: InvoiceType) => state.filter.includes(invoice.status)
        );

        return {
          ...state,
          originalData: action.payload,
          data: { data: filteredData },
        };
      }

      return state;
    case 'addFilter':
      if (!state.filter.includes(action.payload)) {
        state.filter.push(action.payload);
      }

      if (state.data) {
        const filteredData = state.originalData.data.filter(
          (invoice: InvoiceType) => state.filter.includes(invoice.status)
        );

        return {
          ...state,
          filter: state.filter,
          data: { data: filteredData },
        };
      }

      return state;

    case 'removeFilter':
      if (state.filter.includes(action.payload)) {
        state.filter.splice(state.filter.indexOf(action.payload), 1);
      }

      if (state.data) {
        const filteredData = state.originalData.data.filter(
          (invoice: InvoiceType) => state.filter.includes(invoice.status)
        );

        return {
          ...state,
          filter: state.filter,
          data: { data: filteredData },
        };
      }

      return state;
    case 'error':
      return state;
    default:
      throw new Error('Invalid action');
  }
};
