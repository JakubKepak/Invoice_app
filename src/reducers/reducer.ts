// not yet used in the code

type Action =
  | { type: 'loading' }
  | { type: 'error'; payload: any }
  | { type: 'success'; payload: any };

export const reducer = (state: any, action: Action) => {
  switch (action.type) {
    case 'loading':
      return { status: 'loading', data: undefined, error: null };
    case 'error':
      return { status: 'error', data: undefined, error: action.payload };
    case 'success':
      return { status: 'success', data: action.payload, error: null };
    default:
      throw new Error('invalid action');
  }
};
