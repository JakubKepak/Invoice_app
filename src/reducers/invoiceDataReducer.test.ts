import { invoiceDataReducer, initialState } from 'reducers/invoiceDataReducer';

describe('invoiceDataReducer', () => {
  it('should return initial state', () => {
    expect(
      invoiceDataReducer(initialState, { type: 'refresh', payload: undefined })
    ).toMatchSnapshot();
  });
});
