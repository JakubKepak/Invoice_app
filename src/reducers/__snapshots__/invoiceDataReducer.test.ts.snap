// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invoiceDataReducer should return all but paid invoices 1`] = `
Object {
  "data": Object {
    "data": Array [
      Object {
        "clientAddress": Object {
          "city": "Bradford",
          "country": "United Kingdom",
          "postCode": "BD1 9PB",
          "street": "84 Church Way",
        },
        "clientEmail": "alexgrim@mail.com",
        "clientName": "Alex Grim",
        "createdAt": "2021-08-21",
        "description": "Graphic Design",
        "id": "XM9141",
        "items": Array [
          Object {
            "name": "Banner Design",
            "price": 156,
            "quantity": 1,
            "total": 156,
          },
          Object {
            "name": "Email Design",
            "price": 200,
            "quantity": 2,
            "total": 400,
          },
        ],
        "paymentDue": "2021-09-20",
        "paymentTerms": 30,
        "senderAddress": Object {
          "city": "London",
          "country": "United Kingdom",
          "postCode": "E1 3EZ",
          "street": "19 Union Terrace",
        },
        "status": "PENDING",
        "total": 556,
      },
    ],
  },
  "filter": Array [
    "PENDING",
    "DRAFT",
  ],
  "originalData": Object {
    "data": Array [
      Object {
        "clientAddress": Object {
          "city": "Sharrington",
          "country": "United Kingdom",
          "postCode": "NR24 5WQ",
          "street": "106 Kendell Street",
        },
        "clientEmail": "jensenh@mail.com",
        "clientName": "Jensen Huang",
        "createdAt": "2021-08-18",
        "description": "Re-branding",
        "id": "RT3080",
        "items": Array [
          Object {
            "name": "Brand Guidelines",
            "price": 1800.9,
            "quantity": 1,
            "total": 1800.9,
          },
        ],
        "paymentDue": "2021-08-19",
        "paymentTerms": 1,
        "senderAddress": Object {
          "city": "London",
          "country": "United Kingdom",
          "postCode": "E1 3EZ",
          "street": "19 Union Terrace",
        },
        "status": "PAID",
        "total": 1800.9,
      },
      Object {
        "clientAddress": Object {
          "city": "Bradford",
          "country": "United Kingdom",
          "postCode": "BD1 9PB",
          "street": "84 Church Way",
        },
        "clientEmail": "alexgrim@mail.com",
        "clientName": "Alex Grim",
        "createdAt": "2021-08-21",
        "description": "Graphic Design",
        "id": "XM9141",
        "items": Array [
          Object {
            "name": "Banner Design",
            "price": 156,
            "quantity": 1,
            "total": 156,
          },
          Object {
            "name": "Email Design",
            "price": 200,
            "quantity": 2,
            "total": 400,
          },
        ],
        "paymentDue": "2021-09-20",
        "paymentTerms": 30,
        "senderAddress": Object {
          "city": "London",
          "country": "United Kingdom",
          "postCode": "E1 3EZ",
          "street": "19 Union Terrace",
        },
        "status": "PENDING",
        "total": 556,
      },
    ],
  },
}
`;

exports[`invoiceDataReducer should return initial state 1`] = `
Object {
  "data": undefined,
  "filter": Array [
    "PAID",
    "PENDING",
    "DRAFT",
  ],
  "originalData": undefined,
}
`;

exports[`invoiceDataReducer should return state after initial fetch 1`] = `
Object {
  "data": Object {
    "data": Array [
      Object {
        "clientAddress": Object {
          "city": "Sharrington",
          "country": "United Kingdom",
          "postCode": "NR24 5WQ",
          "street": "106 Kendell Street",
        },
        "clientEmail": "jensenh@mail.com",
        "clientName": "Jensen Huang",
        "createdAt": "2021-08-18",
        "description": "Re-branding",
        "id": "RT3080",
        "items": Array [
          Object {
            "name": "Brand Guidelines",
            "price": 1800.9,
            "quantity": 1,
            "total": 1800.9,
          },
        ],
        "paymentDue": "2021-08-19",
        "paymentTerms": 1,
        "senderAddress": Object {
          "city": "London",
          "country": "United Kingdom",
          "postCode": "E1 3EZ",
          "street": "19 Union Terrace",
        },
        "status": "PAID",
        "total": 1800.9,
      },
      Object {
        "clientAddress": Object {
          "city": "Bradford",
          "country": "United Kingdom",
          "postCode": "BD1 9PB",
          "street": "84 Church Way",
        },
        "clientEmail": "alexgrim@mail.com",
        "clientName": "Alex Grim",
        "createdAt": "2021-08-21",
        "description": "Graphic Design",
        "id": "XM9141",
        "items": Array [
          Object {
            "name": "Banner Design",
            "price": 156,
            "quantity": 1,
            "total": 156,
          },
          Object {
            "name": "Email Design",
            "price": 200,
            "quantity": 2,
            "total": 400,
          },
        ],
        "paymentDue": "2021-09-20",
        "paymentTerms": 30,
        "senderAddress": Object {
          "city": "London",
          "country": "United Kingdom",
          "postCode": "E1 3EZ",
          "street": "19 Union Terrace",
        },
        "status": "PENDING",
        "total": 556,
      },
    ],
  },
  "filter": Array [
    "PAID",
    "PENDING",
    "DRAFT",
  ],
  "originalData": Object {
    "data": Array [
      Object {
        "clientAddress": Object {
          "city": "Sharrington",
          "country": "United Kingdom",
          "postCode": "NR24 5WQ",
          "street": "106 Kendell Street",
        },
        "clientEmail": "jensenh@mail.com",
        "clientName": "Jensen Huang",
        "createdAt": "2021-08-18",
        "description": "Re-branding",
        "id": "RT3080",
        "items": Array [
          Object {
            "name": "Brand Guidelines",
            "price": 1800.9,
            "quantity": 1,
            "total": 1800.9,
          },
        ],
        "paymentDue": "2021-08-19",
        "paymentTerms": 1,
        "senderAddress": Object {
          "city": "London",
          "country": "United Kingdom",
          "postCode": "E1 3EZ",
          "street": "19 Union Terrace",
        },
        "status": "PAID",
        "total": 1800.9,
      },
      Object {
        "clientAddress": Object {
          "city": "Bradford",
          "country": "United Kingdom",
          "postCode": "BD1 9PB",
          "street": "84 Church Way",
        },
        "clientEmail": "alexgrim@mail.com",
        "clientName": "Alex Grim",
        "createdAt": "2021-08-21",
        "description": "Graphic Design",
        "id": "XM9141",
        "items": Array [
          Object {
            "name": "Banner Design",
            "price": 156,
            "quantity": 1,
            "total": 156,
          },
          Object {
            "name": "Email Design",
            "price": 200,
            "quantity": 2,
            "total": 400,
          },
        ],
        "paymentDue": "2021-09-20",
        "paymentTerms": 30,
        "senderAddress": Object {
          "city": "London",
          "country": "United Kingdom",
          "postCode": "E1 3EZ",
          "street": "19 Union Terrace",
        },
        "status": "PENDING",
        "total": 556,
      },
    ],
  },
}
`;
